### ISSUE TITLE
[REVIEW]: ldaPrototype: A method in R to get a Prototype of multiple Latent Dirichlet Allocations

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2181]
**Submitting author:** @JonasRieger (<a href="http://orcid.org/0000-0002-0007-4478">Jonas Rieger</a>)
**Repository:** <a href="https://github.com/JonasRieger/ldaPrototype" target ="_blank">https://github.com/JonasRieger/ldaPrototype</a>
**Version:** v0.2.0
**Editor:** @karthik
**Reviewers:** @tommyjones, @bstewart
**Archive:** <a href="https://doi.org/10.5281/zenodo.3945836" target="_blank">10.5281/zenodo.3945836</a>

## Status

[![status](https://joss.theoj.org/papers/ecce89f69453dc2ee0c697fbcb776de8/status.svg)](https://joss.theoj.org/papers/ecce89f69453dc2ee0c697fbcb776de8)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/ecce89f69453dc2ee0c697fbcb776de8"><img src="https://joss.theoj.org/papers/ecce89f69453dc2ee0c697fbcb776de8/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/ecce89f69453dc2ee0c697fbcb776de8/status.svg)](https://joss.theoj.org/papers/ecce89f69453dc2ee0c697fbcb776de8)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/JonasRieger/ldaPrototype" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@tommyjones, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @karthik know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/JonasRieger/ldaPrototype">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@JonasRieger) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/JonasRieger/ldaPrototype">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@JonasRieger) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @tommyjones it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.5281/zenodo.3591068 is OK
- 10.21105/joss.00135 is OK
- 10.18637/jss.v040.i13 is OK
- 10.1073/pnas.0307752101 is OK
- 10.1162/jmlr.2003.3.4-5.993 is OK
- 10.1145/2133806.2133826 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02181/joss.02181/10.21105.joss.02181.pdf)

**Comment by arfon** [comment]
@TommyJones - please carry out your review in this issue by updating the checklist above and giving feedback in this issue. The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html

Any questions/concerns please let @karthik or I know.

**Comment by TommyJones** [comment]
Thanks @arfon! 

The paper and software look great, but there are a couple issues, one is a bit of a blocker to me finishing the review.

The big one:
* There is no example usage of how to use the software for an analysis problem. A great place to put this would be in the README, showing basic usage. If you want to cover more ground than one would typically put in a README, a vignette is a good place. But without this, I'm not sure where to start and thus can't check functionality.

Minor issues:
* The community guidelines do not have a clear statement or instructions for contributing. Adding a sentence to the bottom of the README would fix that right up.
* Functionality documentation is "OK" according to the review criteria. Not all functions have examples, though many do. (Note: as I see it, it's still a "pass" but want to flag that it could be improved.)

One thing that is out of scope for JOSS but I think is worth mentioning: the documentation for parameter inputs to functions is clear but non-standard for R packages. Comes off as a little odd.

**Comment by TommyJones** [comment]
I've opened two issues in the target repo:

[This](https://github.com/JonasRieger/ldaPrototype/issues/1#issue-611517976) is for an example analysis to README or vignette.

[This](https://github.com/JonasRieger/ldaPrototype/issues/2#issue-611518164) is for contribution guidelines.

**Comment by JonasRieger** [comment]
Thanks for the comments. I have improved the points. I will also look again where I can add examples in the documentation of single functions. I deliberately chose this style for the parameter documentation because it is clearer than the standard in R (which is mostly very rudimentary and merely prosaic). It is inspired by the packages [mlr](https://github.com/mlr-org/mlr) and [batchtools](https://github.com/mllg/batchtools).

**Comment by TommyJones** [comment]
Great. I'll give the example a shot this week to test functionality etc.

**Comment by TommyJones** [comment]
@arfon, @karthik, @JonasRieger I've completed my review. Should this close out the issue?

Looks like a useful package! Opened up another issue in the target repo that I think will make for a more pleasant user experience, but think it falls outside of scope for JOSS as I understand it.

**Comment by arfon** [comment]
> @arfon, @karthik, @JonasRieger I've completed my review. Should this close out the issue?

Thanks! Please leave this issue open. As I mentioned in #2156 we still need to add a second reviewer in here before we can consider this review complete.

**Comment by JonasRieger** [comment]
@arfon @karthik In case you missed it, I would like to mention that a second reviewer was found in https://github.com/openjournals/joss-reviews/issues/2156.

**Comment by arfon** [comment]
@whedon add @bstewart as reviewer

**Comment by whedon** [comment]
OK, @bstewart is now a reviewer

**Comment by arfon** [comment]
:wave:  @bstewart - many thanks for agreeing to review this submission - I've just helped out @karthik by adding your checklist to the top of this issue thread.

Please carry out your review in this issue by updating the checklist above and giving feedback in this issue. The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html

Any questions/concerns please let @karthik or myself know.

**Comment by karthik** [comment]
Thanks @arfon!

@bstewart Thanks very much for agreeing to review. Please reach out if you have any questions.

**Comment by karthik** [comment]
*Pinged Brandon today*

**Comment by karthik** [comment]
@bstewart Can you please update us on the status of your review? Can you also let us know if you are unable to complete your review?

**Comment by bstewart** [comment]
I'm really sorry all- should complete this weekend. More soon.

**Comment by karthik** [comment]
Thanks Brandon. Much appreciated! 🙏

**Comment by bstewart** [comment]
Hi all, my apologies for the long delay in my review.  I have posted a few issues in the target repo around the [paper](https://github.com/JonasRieger/ldaPrototype/issues/5), the need to [document defaults](https://github.com/JonasRieger/ldaPrototype/issues/6) and a small issue in the [README](https://github.com/JonasRieger/ldaPrototype/issues/4).

I have a broader philosophical question about the purpose of this package.  The implicit theory put forward by the author is that the best model to use is one that is broadly representative of other solutions.  But LDA is an optimization problem.  The reason we get different answers for different seeds is because that optimization problem is sufficiently hard to solve that we get stuck in local optima (a hopefully accessible introduction is here).  Under the optimization perspective ,the best solution is the one that maximizes the objective function.  So if we have a set of better and worse solutions why would we want one in the middle? This problem is made more vexing by the fact that certain decisions in the underlying `lda` package almost surely make this problem worse.  For example, it isn’t possible to retain anything more than the final draw of the Gibbs sampler for the topic-word counts although results would almost certainly stabilize (at least somewhat!) if you averaged over many draws.  

I think this is an addressable concern.  The central advantage of the proposed procedure from where I sit is a kind of transparency. It suggests at least that of various solutions which are presumably locally optimal, this one shares some similarities with all of them.

Having not reviewed for this journal, I’m not really sure whether or not this last point is really in scope.  The package does what it says it does, but arguably this speaks to the question about the statement of need.  While I think a statement about why this is an attractive strategy in the paper would be preferred, I leave it to the editorial team (@karthik?) to make a judgment about whether such a discussion is relevant for publication here. I'm happy to remove this if it is not the kind of feedback you were looking for.

I'm running some longer tests of the functions which will take another few hours and then will report back on the last few bits.

**Comment by karthik** [comment]
> Having not reviewed for this journal, I’m not really sure whether or not this last point is really in scope

I think this is useful and it would be good for the authors to address it, even briefly in the paper.

**Comment by karthik** [comment]
JonasRieger Can you take a look at Brandon's comments above and address those concerns?

**Comment by bstewart** [comment]
Okay the last thing I wanted to check finished running.  A core claim here is that the estimation is more stable under the proposed procedure.  I wanted to examine the evidence for that claim.  I ran the default dataset for 100 seeds with 100 topics (essentially all default settings).  I then did it again with a different set of seeds.  The question is how similar are the two sets of results?

After aligning the topics, here are the first three.

> Topic 1: 
> said, dlrs, company, mln, pct, stock, inc 
> said, dlrs, company, shares, mln, inc, common 
> Topic 2: 
> computer, technology, software, programmers, chinese, employs, francisco 
> software, programmers, springboard, computer, applications, ever, lisp 
> Topic 3: 
> tower, washington, secretary, last, arms, criticism, state 
> since, state, widespread, commission, made, william, thought 

The first topic seems pretty well aligned, the next two are less so.  The rest of the first 10 topics are at the very bottom of this message (which are all I have examined).  The next relevant question is whether the two prototypes are more similar to each other than any other pair of models.  There is a question about the right measure of similar, so here I consider two, the correlation between the normalized and the unnormalized topic matrix (the latter being a little bit easier because you get some correlation just out of raw word frequencies).

![image](https://user-images.githubusercontent.com/577337/84960539-df579800-b0cf-11ea-8cee-1931cc2c645b.png)

The plot above shows the correlations of the prototypes with a red dashed line and a random sample of pairwise model comparisons as a null distribution.  We can see that while it is definitely more stable than random pairs, it isn't dramatically more stable.

There are plenty of things to criticize about this procedure.  It was a first cut at investigating what I took to be an empirical claim of the paper.  For one thing, this may be simply because the dataset considered here is relatively small and idiosyncratic.  However, evidence like this feels relevant to the long term project.  

Complete code to regenerate these results is at the very bottom.  Note that while the LDA models run quickly the prototyping can take a couple of hours.


> Topic 4: 
 > orders, economy, midland, took, laws, adversely, caused 
 > orders, fed, decline, capital, economists, goods, wyss 
> Topic 5: 
 > said, january, december, year, pct, compared, rose 
 > said, january, compared, december, new, pct, year 
> Topic 6: 
 > network, texas, behalf, comptroller, county, create, guaranteed 
 > unit, capital, sources, power, made, alternate, bear 
> Topic 7: 
 > market, delay, likely, stocks, current, news, well 
 > stock, banking, longer, market, delay, brazilian, equity 
> Topic 8: 
 > said, banks, interest, week, will, approval, terms 
 > said, inc, reuter, first, industries, report, defense 
> Topic 9: 
 > inc, industries, reuter, said, burlington, cofab, petrochemical 
 > interest, banks, bank, reuter, request, shock, afternoon 
> Topic 10: 
 > lion, mortgage, red, expects, partnership, covering, hotels 
 > lion, red, limited, previously, bb, computerrelated, covering 


```
#you will need the clue package for this

#run two default runs on the package with non-overlapping seeds
#we search over 100 solutions for the prototype here
reps_1 = LDARep(docs = reuters_docs, vocab = reuters_vocab,
              n = 100, seeds = 1:100)
reps_2 = LDARep(docs = reuters_docs, vocab = reuters_vocab,
              n = 100, seeds = 101:200)
#we extract the prototypes
p1 <- getPrototype(reps_1)
p2 <- getPrototype(reps_2)
#align the prototypes using the innerproducts of the beta matrix
align_it <- function(m1,m2) {
  x <- tcrossprod(getLDA(m1)$topics,getLDA(m2)$topics)
  clue::solve_LSAP(x, maximum = TRUE)
}
align <- align_it(p1,p2)
#compare the corresponding top words
words1 <- tosca::topWords(getTopics(getLDA(p2)), 7)[,align[1:10]]
words2 <- tosca::topWords(getTopics(getLDA(p1)), 7)[,1:10]
#print solutions
for(i in 1:10) {
  cat(sprintf("> Topic %i: \n > %s \n > %s \n", i, 
              paste0(words1[,i],collapse=", "),
              paste0(words2[,i],collapse=", ")))
}
#check the correlations
b1 <- getTopics(getLDA(p2))[align,]
b2 <- getTopics(getLDA(p1))
c1 <- cor(c(b2),c(b1)) #without normalization
c2 <- cor(c(b2/rowSums(b2)),c(b1/rowSums(b1))) #with normalization



align_it <- function(m1,m2) {
  x <- tcrossprod(m1$topics,m2$topics)
  clue::solve_LSAP(x, maximum = TRUE)
}
set.seed(3253248)
grid <- cbind(sample.int(100,500,replace=TRUE),sample.int(100,500,replace=TRUE))
no_norm <- vector(mode="numeric",length=100)
norm <- vector(mode="numeric",length=100)
for(i in 1:nrow(grid)) {
  j <- grid[i,1]
  k <- grid[i,2]
  align <- align_it(reps_1$lda[[k]],reps_2$lda[[j]])
  b1 <- getTopics(reps_2$lda[[j]])[align,]
  b2 <- getTopics(reps_1$lda[[k]])
  no_norm[i] <- cor(c(b2),c(b1)) #without normalization
  norm[i] <- cor(c(b2/rowSums(b2)),c(b1/rowSums(b1))) #with normalization
}

par(mfrow=c(1,2))
hist(norm,main ="Normalizing",xlab="Rho",xlim=c(.2,.8))
abline(v=c2,lty=2,col="red",lwd=3)
hist(no_norm,main="No Normalizing",xlab="Rho",xlim=c(.2,.8))
abline(v=c1,lty=2,col="red",lwd=3)
```

**Comment by JonasRieger** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02181/joss.02181/10.21105.joss.02181.pdf)

**Comment by JonasRieger** [comment]
Thanks for all the feedback. I updated the default parametrization (no default for K) and fixed the bug in the readMe. I also improved the mentioned points in the paper. I hope this way the contribution will be clearer.

**Comment by bstewart** [comment]
@JonasRieger - No problem, happy to help.
I still disagree with the characterization of two things in the draft.  I don't feel like these things need to change, but let me try to convince you.

1. I still think the language "This contradicts basically good scientific practice" is too strong.  Referencing the latter part of the sentence seems fine to me (i.e. obviously searching to fit your hypothesis is bad), but I see no reason that subjective assessments of model fit are any worse than a measure of model fit based on average similarity to a sample of local posterior modes (your approach).

2.In the Nguyen et al (2014), I don't think the averaging is a weakness.  Honestly, from a Bayesian perspective it is incredibly strange that standard practice isn't to do some kind of averaging.  What does one lose by abandoning a single sample from the posterior?  There is a potential problem with label switching if the chain is run long enough for sure, but honestly that is technically a problem with the single draws as well because the sampler is collapsed.

Anyway I don't feel like either of these should hold you up- just making another attempt to convince you.

@karthik @arfon I think I'm all set unless you need something else from me.

**Comment by JonasRieger** [comment]
The statement actually refers mainly to the method of consciously choosing such models that correspond to the hypothesis. I have changed the specific wording to make this clear. @whedon generate pdf

The weakness of averaging is merely that not all facets of the model can be considered, or, as the case may be, it may become significantly more complex. The token specific assignments cannot be specified concretely, but at most probabilistically. This could lead to problems for some users who perform (partial) analyses on these topic assignments.

I am very interested in comparing the differences of the different selection and averaging algorithms. For this purpose I plan to do a qualitative study in the near future. However, I do not see any method as superior to the others by now.

Thank you again for your feedback and input. In fact, very helpful!

**Comment by JonasRieger** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02181/joss.02181/10.21105.joss.02181.pdf)

**Comment by bstewart** [comment]
Glad you found it helpful.  Re: averaging- I take your point that the assignments are not concrete, but in fairness their concreteness is an illusion.  What you call concrete is just the particular realization of the Gibbs sampler on the draw you happened to stop things.

In my field of study (and most I know of), people are either examining theta (the documents by topics) or beta (words by topics). Both of these have the same form under averaging (values on the simplex).

**Comment by JonasRieger** [comment]
I completely agree with that.

**Comment by JonasRieger** [comment]
short note: I will be on vacation for one week starting saturday 27th.

**Comment by JonasRieger** [comment]
In my opinion, the review should be complete. What should I do next?

**Comment by karthik** [comment]
@JonasRieger Can you please archive your software on Zenodo and post a DOI here please?

**Comment by JonasRieger** [comment]
off course: 10.5281/zenodo.3945836 is the DOI of today's archive/version
10.5281/zenodo.3597978 is the concept DOI

**Comment by karthik** [comment]
@whedon set 10.5281/zenodo.3945836 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3945836" target="_blank">10.5281/zenodo.3945836</a> is the archive.

**Comment by karthik** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02181/joss.02181/10.21105.joss.02181.pdf)

**Comment by karthik** [comment]
Over you to @openjournals/joss-eics

**Comment by arfon** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.5281/zenodo.3591068 is OK
- 10.21105/joss.00135 is OK
- 10.18637/jss.v040.i13 is OK
- 10.21105/joss.00774 is OK
- 10.18637/jss.v091.i02 is OK
- 10.3115/v1/D14-1182 is OK
- 10.1073/pnas.0307752101 is OK
- 10.1162/jmlr.2003.3.4-5.993 is OK
- 10.1145/2133806.2133826 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1572

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1572, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
☝️ @karthik - running `@whedon accept` will add the appropriate label (`recommend-accept`) for the EiC on rotation to pick the paper up (this isn't me this week).

**Comment by JonasRieger** [comment]
I'm sorry. There was a single spelling mistake in one reference, which I have corrected now.

**Comment by labarba** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02181/joss.02181/10.21105.joss.02181.pdf)

**Comment by labarba** [comment]
@karthik @JonasRieger -- the version tag on the repository does not match the version here. Please confirm that we should update here to `0.2.0`.

**Comment by labarba** [comment]
@JonasRieger -- please update the title metadata of your Zenodo deposit to match the JOSS paper title

**Comment by JonasRieger** [comment]
@labarba I confirm that the version here should be updated to `0.2.0`. I also updated the title of the Zenodo deposit.

**Comment by karthik** [comment]
@whedon set v0.2.0 as version

**Comment by whedon** [comment]
OK. v0.2.0 is the version.

**Comment by labarba** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1283669397643702272) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1577
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02181](https://doi.org/10.21105/joss.02181)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by labarba** [comment]
Congratulations, @JonasRieger, your JOSS paper is published! 🚀 

Huge thanks to our editor: @karthik, and reviewers: @TommyJones, @bstewart 🙏

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02181/status.svg)](https://doi.org/10.21105/joss.02181)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02181">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02181/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02181/status.svg
   :target: https://doi.org/10.21105/joss.02181
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02181/status.svg)](https://doi.org/10.21105/joss.02181)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
